aten/src/ATen/core/type.cpp

```diff
@@ -748,6 +752,104 @@ TupleTypePtr TupleType::createNamed(
       field_types, qualName, schema)); // NOLINT(modernize-make-shared)
 }
 
+void flatten_union(TypePtr& type, std::vector<TypePtr>& res) {
+  if (type->kind() == UnionType::Kind) {
+    for (auto inner_type : type->expect<UnionType>()->types()) {
+      flatten_union(inner_type, res);
+    }
+  }
+  else if (type->kind() == OptionalType::Kind) {
+    auto inner_type = type->expect<OptionalType>()->getElementType();
+    if (inner_type->kind() == UnionType::Kind) {
+      flatten_union(inner_type, res);
+    }
+    else {
+      res.emplace_back(inner_type);
+    }
+    res.emplace_back(NoneType::get());
+  }
+  else {
+    res.emplace_back(type);
+  }
+}
+
+UnionType::UnionType(std::vector<TypePtr> types) :
+  Type(TypeKind::UnionType) {
+  // Flatten Unions of Unions while populating `types_`
+  for (auto type : types) {
+    flatten_union(type, types_);
+  }
+  // We want the elements to be sorted so we can easily compare two UnionType
+  // objects for equality in the future
+  std::sort(types_.begin(), types_.end(),
+  [](const TypePtr a, const TypePtr b) -> bool {
+    return a->kind() > b->kind();
+  });
+  // Filter out duplicate types
+  types_.erase(std::unique(types_.begin(), types_.end()), types_.end());
+  // Check if this Union allows None
+  can_hold_none_ = std::any_of(types_.begin(), types_.end(),
+    [](const TypePtr type) { return type->kind() == NoneType::Kind; });
+}
+
+bool UnionType::isSubtypeOfExt(const TypePtr& rhs_, std::ostream* why_not) const {
+  if (auto union_rhs = rhs_->cast<UnionType>()) {
+    // Fast path: The Union `types` vector is always sorted, so we can
+    // do an O(n) linear comparison of both vectors
+    if (this->types() == union_rhs->types()) {
+      return true;
+    }
+    // Final O(n^2) check since the vector comparison will falsely
+    // flag subtypes as not equal
+    return std::all_of(this->types_.begin(), this->types_.end(),
+      [&](TypePtr this_type) -> bool {
+        return union_rhs->can_hold_type(this_type);
+    });
+  }
+  if (auto optional_rhs = rhs_->cast<OptionalType>()) {
+    return types_.size() == 2
+      && can_hold_type(optional_rhs->getElementType())
+      && can_hold_none();
+  }
+  if (Type::isSubtypeOfExt(rhs_, why_not)) {
+    return true;
+  }
+  return false;
+}
+
+std::string UnionType::str() const {
+  std::stringstream ss;
+  ss << "Union[";
+  for (size_t i = 0; i < types().size(); ++i) {
+    if (i > 0) {
+      ss << ", ";
+    }
+    ss << types()[i]->str();
+  }
+  ss << "]";
+  return ss.str();
+}
+
+  // Check if a given TypePtr is an allowable member of this Union
+  bool UnionType::can_hold_type(TypePtr& type) const {
```

Comment:
Why do we need const and non const version ?

Reply:
There are lots of places where the `const`-ness is already set in stone. I didn't think it made sense to trace back all those calls and change them all to be non-`const` so that they could work with Union, since that could have implications for the larger codebase. I might have underestimated the importance of keeping the API small, though. What would you suggest?

Reply:
I guess i figured we just needed a const version, but it's not a big deal